test/system/src/test/java/io/pravega/test/system/MultiReaderWriterTxnWithFailoverTest.java

```diff
@@ -0,0 +1,444 @@
+
+        currentWriteCount1 = eventWriteCount.get();
+        currentReadCount1 = eventReadCount.get();
+
+        log.info("Stop write flag status {}", stopWriteFlag);
+        log.info("Stop read flag status {}", stopReadFlag);
+        log.info("Event data {}", eventData.get());
+        log.info("Read count {}", eventReadCount.get());
+        log.info("Read count with SSS and controller failover after sleep {}", currentReadCount1);
+        log.info("Write count with SSS and controller failover after sleep {} ", currentWriteCount1);
+    }
+
+    private CompletableFuture<Void> startWritingIntoTxn(final AtomicLong data, final EventStreamWriter<Long> writer,
+                                                 final AtomicBoolean stopWriteFlag, final AtomicLong eventWriteCount) {
+        return CompletableFuture.runAsync(() -> {
+            while (!stopWriteFlag.get()) {
+                Transaction<Long> transaction = null;
+                    try {
+                        transaction = Retry.withExpBackoff(10, 10, 20, ofSeconds(1).toMillis())
+                                .retryingOn(MultiReaderWriterTxnWithFailoverTest.TxnCreationFailedException.class)
+                                .throwingOn(RuntimeException.class)
+                                .run(() -> createTransaction(writer, stopWriteFlag));
+
+                        //each transaction has 50 events
+                        for (int j = 0; j < 50; j++) {
+                            long value = data.incrementAndGet();
+                            Thread.sleep(100);
+                            transaction.writeEvent(String.valueOf(value), value);
+                            log.debug("Writing event {} into transaction {}", value, transaction);
+                        }
+                        //commit Txn
+                        transaction.commit();
+                        eventWriteCount.getAndAdd(50);
+                        log.debug("Transaction {} committed", transaction);
+                        log.debug("Transaction status {}", transaction.checkStatus());
+                    } catch (Throwable e) {
+                        log.warn("Exception while writing events in the transaction : {}", e);
+                        log.debug("transaction with id {}  failed", transaction.getTxnId());
+                    }
+            }
+        }, executorService);
+    }
+
+    private Transaction<Long> createTransaction(EventStreamWriter<Long> writer, final AtomicBoolean exitFlag) {
+        Transaction<Long> txn = null;
+        try {
+            //Default max scale grace period is 30000
+            txn = writer.beginTxn(5000, 3600000, 29000);
+            log.info("Transaction created with id:{} ", txn.getTxnId());
+        } catch (RuntimeException ex) {
+            log.info("Exception encountered while trying to begin Transaction ", ex.getCause());
+            final Class<? extends Throwable> exceptionClass = ex.getCause().getClass();
```

Comment:
instanceof is far cheaper than getClass() and it is cleaner because it can be done inline in the if and handles subclasses.